### PR TITLE
Add Netflix DGS Codegen plugin support

### PR DESCRIPTION
**Pull request by bclozel**
This commit adds a new dependency to start.spring.io, the "Netflix Data
Graph Service Codegen". This will not add a dependency to generated
project, but rather configure a Maven or a Gradle plugin in the build.

This Codegen plugin parses GraphQL schema files describing remote
GraphQL APIs and generates Java classes to easily create queries and
deserialize GraphQL responses.

### CODE REVIEW

File: start-site/src/main/resources/application.yml [hunk 1]
```diff
@@ -203,6 +203,16 @@ initializr:
           artifactId: spring-boot
           description: Support for compiling Spring applications to native executables using the GraalVM native-image compiler.
           starter: false
+        - name: GraphQL DGS Code Generation
+          id: dgs-codegen
+          compatibilityRange: "3.0.0-M1"
+          groupId: com.netflix.graphql.dgs.codegen
+          artifactId: graphql-dgs-codegen-gradle
+          mappings:
```
Comment (snicoll) [inline start-site/src/main/resources/application.yml:211]: Wondering if we need a mapping here considering that the artifact is just a trigger for the build system setup.

File: start-site/src/test/java/io/spring/start/site/extension/dependency/dgs/DgsCodegenHelpDocumentCustomizerTests.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+class DgsCodegenHelpDocumentCustomizerTests extends AbstractExtensionTests {
+
+	@Autowired
+	private MustacheTemplateRenderer templateRenderer;
+
+	@Test
+	void mavenBuildAddsLinkToMavenCodegenPlugin() {
+
+		assertHelpDocument("maven-project", "dgs-codegen").contains("## GraphQL code generation with DGS")
+			.containsSubsequence("This project has been configured to use the Netflix DGS Codegen plugin.",
```
Comment (snicoll) [inline start-site/src/test/java/io/spring/start/site/extension/dependency/dgs/DgsCodegenHelpDocumentCustomizerTests.java:42]: I wonder if that's strictly necessary as any wording change needs the test to be updated. I think we're probably a bit more relaxed for other help document tests. Checking that the proper link according to the build system is specified (and the non matching one is not present) looks the real important bit for me.

File: start-site/src/test/java/io/spring/start/site/extension/dependency/dgs/DgsCodegenProjectGenerationConfigurationTests.java [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+
+/**
+ * Tests for {@link DgsCodegenProjectGenerationConfiguration}.
+ *
+ * @author Brian Clozel
+ */
+class DgsCodegenProjectGenerationConfigurationTests extends AbstractExtensionTests {
+
+	@Test
+	void gradleBuildWithoutNativeDoesNotConfigureNativeBuildTools() {
```
Comment (snicoll) [inline start-site/src/test/java/io/spring/start/site/extension/dependency/dgs/DgsCodegenProjectGenerationConfigurationTests.java:33]: Typo as it refers to native?

File: start-site/src/test/java/io/spring/start/site/extension/dependency/dgs/DgsCodegenProjectGenerationConfigurationTests.java [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+		ProjectRequest request = createProjectRequest("web");
+		assertThat(gradleBuild(request)).doesNotContain("com.netflix.dgs.codegen");
+	}
+
+	@Test
+	void mavenBuildWithoutNativeDoesNotConfigureNativeBuildTools() {
```
Comment (snicoll) [inline start-site/src/test/java/io/spring/start/site/extension/dependency/dgs/DgsCodegenProjectGenerationConfigurationTests.java:39]: Typo as it refers to native?

File: start-site/src/main/java/io/spring/start/site/extension/dependency/dgs/DgsCodegenMavenBuildCustomizer.java [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+									.configuration((configuration) -> configuration
+										.add("schemaPaths",
+												(schemaPaths) -> schemaPaths.add("param",
+														"src/main/resources/graphql-client"))
+										.add("packageName", this.packageName + ".codegen")
+										.add("generateDataTypes", "false")
+										.add("addGeneratedAnnotation", "true"))));
+		build.plugins()
+			.add("org.codehaus.mojo", "build-helper-maven-plugin",
+					(plugin) -> plugin.execution("generate-sources", (execution) -> execution.goal("add-source")
```
Comment (snicoll) [inline start-site/src/main/java/io/spring/start/site/extension/dependency/dgs/DgsCodegenMavenBuildCustomizer.java:56]: I wonder if `generate-sources` is a suitable `id` for this. I think `add-dgs-source` or something could be better?

**Review by snicoll — COMMENTED** [review]
this look complete and quite polished already. I've just added a few nits for your consideration. Thanks!

File: start-site/src/main/resources/application.yml (hunk 1 above)
Comment (bclozel) [reply start-site/src/main/resources/application.yml:211]: Right now the only added benefit is that if we upgrade the plugin version and it's not fully published on Maven Central, our verification tests will catch it. I'd be tempted to leave that in for this reason.

File: start-site/src/test/java/io/spring/start/site/extension/dependency/dgs/DgsCodegenHelpDocumentCustomizerTests.java (hunk 2 above)
Comment (bclozel) [reply start-site/src/test/java/io/spring/start/site/extension/dependency/dgs/DgsCodegenHelpDocumentCustomizerTests.java]: I've relaxed the tests, thanks!

File: start-site/src/main/java/io/spring/start/site/extension/dependency/dgs/DgsCodegenMavenBuildCustomizer.java (hunk 5 above)
Comment (bclozel) [reply start-site/src/main/java/io/spring/start/site/extension/dependency/dgs/DgsCodegenMavenBuildCustomizer.java]: Changed the code generation plugin configuration as well.

**Comment by bclozel** [comment]
This PR is still in draft mode, we're discussing some details here and maybe we should link to Spring for GraphQL's docs once https://github.com/spring-projects/spring-graphql/issues/847 is resolved.

**Pull request merged**